### ISSUE TITLE
Complete DynamoDB backend and add integration tests

### ISSUE THREAD

**Issue by JWCook** [JWCook/aiohttp-client-cache#9]
The current state is compatible with the `BaseCache` API, but missing a couple methods, does not run asynchronously, and is untested.

**Note:** `boto3` does not have async support. [aioboto3](https://aioboto3.readthedocs.io/en/latest/usage.html) may be worth looking into.

**Comment by akursar** [comment]
We use dynamoDB as a cache in aws lambda functions that aren't running in a managed VPC, where other options like elasticache aren't easily available. One of our observations is that dynamoDB is pretty slow, usually taking ~10ms to read. So, we are interested in minimizing round trips. Specifically, we'd be looking to:
* skip "contains" searches in the backend's get_response call. I think we'd want to go straight to using read here
* skip "delete" - rely on setting TTL when writing and lean on dynamodb's expiry and that fact that we'll likely overwrite this key if it's expired near immediately
* skip updating last_used - this is interesting but not worth it to us right now

I can put some of this in a DynamoDb-specific Backend, or impl some of the functions like contains/delete to be no-ops in the Cache, but curious if you think any of these concerns are broad enough to make more generic configurations in the base Backend or something.

**Comment by JWCook** [comment]
That all sounds good to me.

`contains()`: Do you still want to cache redirects? If so, you could modify this section in `get_response()`: https://github.com/JWCook/aiohttp-client-cache/blob/eb97971fdc7e2d1e0a7af1554c16453e1d28b691/aiohttp_client_cache/backends/base.py#L164-L166
to:
```python
response = await self.responses.read(key)
if not response:
    redirect_key = str(await self.redirects.read(key))
    response = await self.responses.read(redirect_key)
```
So it would only be 1 call if the response is cached (but still 3 calls for redirects).

`delete()`: That would be great. The same would potentially apply to Redis. You could skip that call with minimal code duplication by adding a class variable, say `has_ttl`, and changing [line #174](https://github.com/JWCook/aiohttp-client-cache/blob/main/aiohttp_client_cache/backends/base.py#L174) to `if not self.has_tty and self.is_cacheable(response):`.

`last_used`: Good point. I added that for the possibility of having an LRU-style cache, but that won't be anytime soon, so I went ahead and disabled that update by default.

**Comment by JWCook** [comment]
Note on tests: you should be able to basically copy [test_redis_backend.py](https://github.com/JWCook/aiohttp-client-cache/blob/main/test/integration/test_redis_backend.py) as a starting point, add any DynamoDB-specific bits that you want, and uncomment [line #63 in build.yml](https://github.com/JWCook/aiohttp-client-cache/tree/main/.github/workflows/build.yml#L63) to run a DynamoDB container in GitHub Actions. I had already added a container to `docker-compose.yml` for local testing, so feel free to modify that as well.

**Comment by JWCook** [comment]
Also, just curious, have you looked into using [DAX](https://aws.amazon.com/dynamodb/dax/) to speed up your queries?

**Comment by akursar** [comment]
> Also, just curious, have you looked into using [DAX](https://aws.amazon.com/dynamodb/dax/) to speed up your queries?

Not really. Just enough to realize it would cost more money. We're looking to move off Lambda to long-lived containers on k8s, and that's driving the adoption of asyncio. We're looking to optimize throughput and cost, not latency. I actually only need the dynamodb support to ease our migration and expect to switch to redis/elasticache once that's more accessible to us.

**Comment by JWCook** [comment]
Gotcha. You're probably already aware, but the recently added [Lambda container integration](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-container-image-support/) is also a pretty attractive option for some container-based workloads. My team has been looking into moving to that for some cases where we'd otherwise use Fargate or ECS. K8s may be the better option in your case, though.

**Comment by akursar** [comment]
Most of our time (90%) is spent waiting on IO. The lambda model where we pay for every 100ms doesn't work very well when we spend 900ms waiting on various IO before responding back to the client with results in 1s :)
We're looking to get better mileage with asyncio and long-lived containers where we can likely support the same activity with less/cheaper compute resources.

**Comment by akursar** [comment]
> That all sounds good to me.
> 
> `contains()`: Do you still want to cache redirects? If so, you could modify this section in `get_response()`:
> 
> https://github.com/JWCook/aiohttp-client-cache/blob/eb97971fdc7e2d1e0a7af1554c16453e1d28b691/aiohttp_client_cache/backends/base.py#L164-L166
> 
> 
> to:
> ```python
> response = await self.responses.read(key)
> if not response:
>     redirect_key = str(await self.redirects.read(key))
>     response = await self.responses.read(redirect_key)
> ```
> 
> So it would only be 1 call if the response is cached (but still 3 calls for redirects).

Right, I don't need support for redirects and so my custom backend/cache doesn't reference them at all. It just has the one cache (dynamodb) for which i have an abbreviated:

    async def get_response(self, key: str) -> Optional[CachedResponse]:
        """This differs from the default impl in the following ways:
           * We do not update 'last_used' ( this is no longer done in main anyway)
           * We are not supporting redirects ( because we don't expect them and are avoiding the complexity
           * We never call 'delete' because we rely on DynamoDB's ttl
        """

        response = await self.responses.read(key)
        if not response:
            return None

        if not self.is_cacheable(response):
            return None

        return response

I was going to ask if we could make redirects an optional feature in the default backend and so only conditionally use them in get_resp / wherever else they're used so that my omission of them isn't so glaring. But I also don't really understand why they're treated differently. That is, why isn't caching redirects just a matter of adding 301 to `allowed_codes`? I'll admit I'm likely hindered by ignorance here. I see some handling of "redirect history" here that goes over my head. But I guess I want to treat 301 no different than 200 or 404. We'll either catch it, or not, but not have separate caches for diff resp codes.
This is creating friction for me for tackling the dynamodb task because I get mentally stuck on "I'm going to have one dynamodb table but the lib expects two."

**Comment by JWCook** [comment]
Say you hit a URL and get 3 redirects in a row followed by the content you actually want to get to. If you cache 301s, what you get back from the cache is the content of the 301 response. So your cache would look something like (simplified):
```python
responses = {
    'https://x.com/redir_1': 'Response <301>: Location: https://x.com/redir_2',
    'https://x.com/redir_2': 'Response <301>: Location: https://x.com/redir_3',
    'https://x.com/redir_3': 'Response <301>: Location: https://x.com/actual_content':,
    'https://x.com/actual_content': 'Response <200> ...',
}
```

But even if you set `allowed_codes=(200, 301)`, those wouldn't actually make it into the responses cache since `aiohttp` follows redirects automatically. Those steps are returned in `ClientResponse.history`. Reference: https://docs.aiohttp.org/en/stable/client_advanced.html#redirection-history

By tracking responses in a separate table, when you hit `https://x.com/redir_1` again, you'll get the final 200 response content right away rather than the 301 response content. That being said, I ported this implementation over from [requests-cache](https://github.com/reclosedev/requests-cache) (which I am also working on at the moment), and there very well could be a better implementation of this. Off the top of my head I can't think of a good way to store everything in the same table, without duplicating response data, and in a way that would work with all backends.

I'm sure there are plenty of cases where you won't expect to encounter many redirects, or otherwise the performance cost of the extra check for redirects isn't worth it. So I think making redirect caching optional in the base class is a great idea. Maybe add an init parameter `cache_redirects` and default it to `True`.

**Comment by akursar** [comment]
Thanks for the info. It doesn't quite make sense to me yet but I'll set up some tests and observe. When you say, "without duplicating response data" - is that because you want to store a single resp blob and return that whether the client hits https://x.com/redir_2 or https://x.com/actual_content in your example? I think if that's your acceptance criteria, it makes much more sense to me. If follow-redirs was on, I'd expect to get the final result for all those diff paths, but for diff keys. So yes there's be duplicate resp data stored, but we'd only have to do one cache read instead of two, which is what I'm looking for. I can see how you might want to optimize one way over another though.

**Comment by JWCook** [comment]
Exactly, it's a storage space vs. latency tradeoff. For, say, a local SQLite db or a Redis instance on the same network, the response time is almost nothing. And for some use cases with lots of large responses, if we duplicate response data, we could be talking about gigabytes of added size. Which in itself could either be a non-issue or a headache, depending on the environment.

Since in your case the response time is more significant and the cost of storage is negligible, this is actually really helpful for thinking through other options. If it wouldn't add too much code complexity, I agree it would be handy to add the option to store everything in one table so your `get_response()` would always result in only one db operation (assuming you're also using TTL and never calling `delete()`). That should be a separate issue and a separate PR, though.

**Comment by akursar** [comment]
Crystal clear, thanks!
